server/src/main/java/org/opensearch/common/logging/DeprecatedMessage.java

```diff
@@ -77,6 +82,9 @@ private static Map<String, Object> fieldMap(String key, String xOpaqueId) {
     }
 
     public boolean isAlreadyLogged() {
-        return !keys.add(keyWithXOpaqueId);
+        if (keyDedupeCache.contains(keyWithXOpaqueId)) {
+            return true;
+        }
+        return keyDedupeCache.size() < MAX_DEDUPE_CACHE_ENTRIES && !keyDedupeCache.add(keyWithXOpaqueId);
```

Comment:
It would result in logging each message when the cache is full. It can stop logging as well when the cache is full.

Reply:
Yes, this will log all deprecation warnings for all new clients after the max size is exceeded. Are you suggesting to invert this logic to stop logging in that case?

Reply:
To explicitly spell out the choices, once the max size is exceeded:

- fail open: log all messages, to ensure the operator knows that 1) they are using deprecated functions, and 2) there are misconfigured clients somewhere misusing the `x-opaque-id` header. The risk is this introduces a performance degradation due to excessive logging.
- fail closed: stop logging all deprecation messages for new clients. This could potentially hide deprecation warnings, but in order to get to this point you'll have historical logs full of deprecation messages that you're probably not looking at anyway.

Reply:
yes, I would prefer ```fail closed``` to prevent impacting OpenSearch performance as this is the reason we introduced the deprecation logging optimization in the first place. Like you said, user would have tons of historic logs full of deprecation messages they can look at them.